crates/ty_python_semantic/src/semantic_index/builder.rs

```diff
@@ -458,13 +490,165 @@ impl<'db, 'ast> SemanticIndexBuilder<'db, 'ast> {
+                        if nested_symbol.is_bound() {
+                            self.place_tables[popped_scope_id].add_nested_scope_with_binding(
+                                resolution.symbol_id,
+                                nested_scope_id,
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        if popped_scope_id.is_global() {
+            // If we've popped the global/module scope, any remaining free variables are
+            // unresolved. The common case for these is built-ins like `print`, and rarer cases are
+            // things like direct insertions into `globals()`. However, if any `nonlocal` free
+            // variables are still unresolved, that's another syntax error.
+            debug_assert!(self.scope_stack.is_empty());
+            for (name, variables) in &popped_free_variables {
+                for variable in variables {
+                    if let Some(nonlocal_identifier) = variable.nonlocal_identifier {
+                        self.report_semantic_error(SemanticSyntaxError {
+                            kind: SemanticSyntaxErrorKind::NoBindingForNonlocal(
+                                name.clone().into(),
+                            ),
+                            range: nonlocal_identifier.range(),
+                            python_version: self.python_version,
+                        });
+                    }
+                }
+            }
+        } else {
+            // Otherwise, add any still-unresolved free variables from nested scopes to the parent
+            // scope's collection, and walk the popped scope's symbol table to collect any new free
+            // variables. During that walk, also record references to global variables.
+            let parent_free_variables = &mut self
+                .scope_stack
+                .last_mut() // current_scope_info_mut() would be a borrock error here
+                .expect("this is not the global/module scope")
+                .free_variables;
+            for (name, variables) in popped_free_variables {
+                parent_free_variables
+                    .entry(name)
+                    .or_default()
+                    .extend(variables);
+            }
+            let popped_place_table = &self.place_tables[popped_scope_id];
+            let mut bound_global_symbols = Vec::new();
+            for symbol in popped_place_table.symbols() {
+                // Collect new implicit (not `nonlocal`) free variables.
+                //
+                // NOTE: Because these variables aren't bound (won't wind up in
+                // `nested_scopes_with_bindings`) and aren't `nonlocal` (can't trigger `nonlocal`
+                // syntax errors), collecting them currently has no effect. We could consider
+                // removing this bit and renaming `free_variables` to say `unresolved_nonlocals`?
```

Comment:
If I delete this block, is there a good way to rerun the CodSpeed instrumentation and see if it helps?